v2/internal/internalguantlettest/gauntlet_test.go

```diff
@@ -0,0 +1,170 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internalgauntlettest
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpchttp"
+	"go.uber.org/yarpc/v2/yarpcrandpeer"
+	"go.uber.org/yarpc/v2/yarpcrouter"
+)
+
+type lifecycle interface {
+	Start(context.Context) error
+	Stop(context.Context) error
+}
+
+func newProcedures() []yarpc.TransportProcedure {
+	return jsonProcedures()
+}
+
+func newInbound(t *testing.T, transport string, listener net.Listener, procedures []yarpc.TransportProcedure) (stop func() error) {
+	var inbound lifecycle
+	router := yarpcrouter.NewMapRouter(_service, procedures)
+
+	switch transport {
+	case _http:
+		inbound = &yarpchttp.Inbound{
+			Listener: listener,
+			Router:   router,
+		}
+
+	default:
+		t.Fatalf("unsupported transport: %q", transport)
+	}
+
+	require.NoError(t, inbound.Start(context.Background()))
+	return func() error { return inbound.Stop(context.Background()) }
+}
+
+// returns a yarpc.Chooser with ID added to the backing peer list
+func newChooser(t *testing.T, chooser string, dialer yarpc.Dialer, id yarpc.Identifier) yarpc.Chooser {
+	update := yarpc.ListUpdates{Additions: []yarpc.Identifier{id}}
+
+	switch chooser {
+	case _random:
+		pl := yarpcrandpeer.New(dialer)
+		pl.Update(update)
+		return pl
+
+	default:
+		t.Fatalf("unsupported peer chooser: %q", chooser)
+	}
+
+	return nil
+}
+
+// returns UnaryOutbounds configured with every specified chooser
+// eg HTTP+round-robin, HTTP+random
+func newOutbounds(t *testing.T, transport string, addr string, choosers []string) (_ []yarpc.UnaryOutbound, stop func() error) {
+	var (
+		outbounds []yarpc.UnaryOutbound
+		dialers   []lifecycle
+	)
+
+	id := yarpc.Address(addr)
+
+	for _, chooser := range choosers {
+		switch transport {
+		case _http:
+			dialer := &yarpchttp.Dialer{}
+			require.NoError(t, dialer.Start(context.Background()))
+			dialers = append(dialers, dialer)
```

Comment:
nit: If therre is always a 1:1 correspondence between `dialers` `outbounds`, and `choosers`, we could allocate their length up-front and directly index into their slices rather than using `append`.